File: google_groups_auth.py
#!/usr/bin/python3

# ============================== SUMMARY =====================================
#
# Program : google_groups_auth.py
# Version : 0.1 ( first beta )
# Date    : May 22, 2018
# Author  : Marcin Kowalczuk - marcin.kowalczuk(at)rtbhouse.com
# Licence : GPL - summary below, full text at http://www.fsf.org/licenses/gpl.txt
#

import sys
import time
import json
import base64
from google.oauth2 import service_account
import googleapiclient.discovery
from pymemcache.client import base
import config_file as config


#---------------------------------------------------------------------------------
def ask_cache(key):
    try:
        log("cache  -  ask key: " + key)
        client = base.Client((config.MemcacheServer, config.MemcachePort))
        response = client.get(key)
    except:
        log("cache  -  problem with memcache, return: None")
        return None

    return response

#---------------------------------------------------------------------------------
def save_cache(key, value):
    try:
        log("cache  -  save key: " + key)
        client = base.Client((config.MemcacheServer, config.MemcachePort))
        # 5 min expiration time
        client.set(key, value, config.MemcacheExpireTime)
    except:
        log("cache  -  problem with saving cache")

#---------------------------------------------------------------------------------
def log(string):
    ''' to improve, very poor logging '''

    f.write(time.ctime() + "  -  " + string + "\n")
    f.flush()

#---------------------------------------------------------------------------------
def get_google_token():
    ''' to improve: use existing token, at the moment we always create new '''

    log("get google token")
    credentials = service_account.Credentials.from_service_account_file(config.ServiceAccountFile, scopes=config.GoogleScopes).with_subject(config.SubjectAccount)
    directory = googleapiclient.discovery.build('admin', 'directory_v1', credentials=credentials)
    return directory

#---------------------------------------------------------------------------------
def ask_google(key):
    ''' fetching user google groups '''

    try:
        directory = get_google_token()
        results = directory.groups().list(userKey=key).execute()
        grupy = json.loads(json.dumps(results))
        email_list = [element['email'] for element in grupy['groups']]
    except:
        email_list = "empty"
        return(email_list)
    
    return(email_list)

#---------------------------------------------------------------------------------
def get_json(user):

    gr = ask_cache(user) 
    if gr is None:
        gr = ask_google(user)
        log("json  -  " + user + " from google")
        save_cache(user, base64.b64encode(json.dumps(gr).encode('utf-8')))
        return base64.b64encode(json.dumps(gr).encode('utf-8'))
    else:
        log("json  -  " + user + " from cache fetched")
        return gr


#---------------------------------------------------------------------------------
def hasMember(user, group):
    ''' checks cache and (if nessesary) ask google api:https://www.googleapis.com/admin/directory/v1/groups/groupKey/hasMember/memberKey '''


    # join user and group to creating uniq key for memcache
    key = user + "#" + group
    isMember = ask_cache(key)
    if isMember is None:
        #ask goole
        try:
            log("auth  -  connect to google /hasMember/")
            directory = get_google_token()
            results = directory.members().hasMember(groupKey=group, memberKey=user).execute()
            isMember = json.loads(json.dumps(results))['isMember']
        except:
            # if error return no (for example group don't exist )
            log("auth  -  error connect to google /hasMember/")
            isMember = False
            
        if isMember is True:
            save_cache(key, "yes")
            return "yes"
        if isMember is False:
            save_cache(key, "no")
            return "no"
        else:
            log("auth  -  error, unsupported google answer /hasMember/, return no")
            return "no"

    else:
        return isMember.decode('utf-8')



#---------------------------------------------------------------------------------
def check_access_by_member(user, authgroups_string):
    ''' Checks whether the given user is a member of the group. Membership can be direct or nested. '''
    authgroups = authgroups_string.split(";")

    # for authgroups member user membership
    for authgroup in authgroups:
        if hasMember(user, authgroup) == "yes":
            log("auth  -  isMember " + user + "->" + authgroup + " -> return yes")
            return "yes"
        else:
            log("auth  -  isMember " + user + "->" + authgroup +" -> no")
            

    log("auth  -  isMember " + user + "->" + authgroup +" -> return no")
    return "no"


#---------------------------------------------------------------------------------
def main():
    request_count = 0
    global f
    f = open(config.LogFile, 'a')  
    while True:
        # Fetch one line from stdin
        request = input()
        request_count += 1
        log("#################   request count: " + str(request_count))
        log("#################   request -  " + request)
        
        # select script function: fetch user grups or checks permissions
        answer = None
        if request.startswith("json"): 
            args = request.split("#")
            # 0:json  1:user
            answer =  get_json(args[1])
            if not config.isBase64Encrypt:
                   answer =  base64.b64decode(answer)
            log("json  -  STDOUT  -  " + answer.decode('utf-8'))
            print(answer.decode('utf-8'))
            continue


        if request.startswith("auth"): 
            args = request.split("#")
            # 0:auth  1:user  2: authgroups
            if args[2] == "":
                #workaround, some requests are empty, 
                #!!!! (resolved, can't use SetEnv in apache conf)
                answer = "think"
                log("auth  -  STDOUT  -  " + args[1]  + " ...think.....")
                print(answer)
                continue
            else:
                #answer =  check_access(args[1], args[2])
                answer =  check_access_by_member(args[1], args[2])
                log("auth  -  STDOUT  -  " + answer)
                print(answer)
                continue

        if answer == None: 
            log("error  -  bad request, STDOUT: NULL")
            print("NULL")

    f.close()
#---------------------------------------------------------------------------------
if __name__ == "__main__":
    main()
#---------------------------------------------------------------------------------

File: Go/google_groups_auth.go
// ============================== SUMMARY =====================================
//
// Program : google_groups_auth.go
// Version : 1.0.go 
// Date    : 05.07.2018
// Author  : Marcin Kowalczuk - marcin.kowalczuk(at)rtbhouse.com
// Licence : GPL - summary below, full text at http://www.fsf.org/licenses/gpl.txt
//
//------------------------------------------------------------------------------------
package main

import (
	"fmt"
	"os"
	"bufio"
	"encoding/json"
	"encoding/base64"
	"strings"
	"log"
	"golang.org/x/net/context"
	"google.golang.org/api/admin/directory/v1"
	"io/ioutil"
	"golang.org/x/oauth2/google"
	"errors"
	"github.com/patrickmn/go-cache"
	"time"
)
//##############################################################################

type Configuration struct {
	GoogleScope	string
	ServiceAccountFile	string
	SubjectAccount	string
	LogFile	string
	Base64Encrypt	bool
}

//##############################################################################
func hasMember(user, group string) bool {

	var key string
	key = user + "#" + group
	log.Printf("CACHE - ask for AUTH key: %s", key)
	val, found := c.Get(key)
	if found{
		log.Printf("CACHE - Found key: %s in cache", key)
		//debug*** //fmt.Printf("z cache ------ %B\n",val)
		return val.(bool)
	}else{
		// ask google
		log.Printf("CACHE - Do not found key: %s in cache\n", key)
		log.Printf("AUTH  - connect to google /has/Member\n")
		directory := get_google_token()

		results, err := directory.Members.HasMember(group,user).Do()
		if err != nil {
          log.Printf("error asking google: %v\n", err)
	}
	log.Printf("CACHE - save response to cache - key: %s\n", key )
		c.Set(key,results.IsMember, cache.DefaultExpiration)
		return results.IsMember
	}

}
//##############################################################################
func check_access_by_member(user, authgroups_string string) (string, error) {
	// Checks whether the given user is a member of the group. Membership can be direct or nested.

	authgroups := strings.Split(authgroups_string, ";")
	//delete \n in last group
	authgroups[len(authgroups)-1] = strings.TrimSuffix(authgroups[len(authgroups)-1],"\n")

	// loop on authgroups
	for _, authgroup := range authgroups{
		if hasMember(user, authgroup){
			log.Printf("AUTH  -  %s isMember %s ? ...... return yes", user, authgroup)
			return "yes", nil
		}else{
			log.Printf("AUTH  - %s is Member %s ? ....... return no", user, authgroup)
		}
	}
	log.Printf("AUTH  - %s is not member of any groups return no", user)
	return "no", nil
}
//##############################################################################
func get_google_token() *admin.Service {
	jsonCredentials, err := ioutil.ReadFile(conf.ServiceAccountFile)
    if err != nil {
        log.Printf("error reading credentials from file: %v", err)
    }

    config, err := google.JWTConfigFromJSON(jsonCredentials, conf.GoogleScope)

    if err != nil {
            log.Printf("Unable to parse client secret file to config: %v", err)
    }

    config.Subject = conf.SubjectAccount
    ctx := context.Background()
    client := config.Client(ctx)
	srv, err := admin.New(client)

    if err != nil {
    	log.Printf("Unable to create directory service %v", err)
     }
	return srv


	}
//##############################################################################
func get_json(user string) (string, error)  {

	// TODO: ask cache
	log.Printf("CACHE - ask for JSON key: %s", user)
	val, found := c.Get(user)
	if found{
		log.Printf("CACHE - Found JSON key: %s in cache", user)
		//debug*** // fmt.Printf("z cache ------ %s\n",val.(string))
		return val.(string), nil
	}else {

		var wyniki []string

		// ask google
		log.Printf("CACHE - Do not found json for user %s \n", user)
		log.Printf("JSON  - connect to google \n")
		directory := get_google_token()

		results, err := directory.Groups.List().UserKey(strings.TrimSuffix(user, "\n")).Do()
		if err != nil {
			log.Printf("error asking google: %v\n", err)
			return "no", errors.New("Problem with asking google")

		}

		if len(results.Groups) == 0 {
			fmt.Print("No groups found.\n")
		} else {
			for _, u := range results.Groups {
				wyniki = append(wyniki, u.Email)
			}
		}
		jsonwynik, _ := json.Marshal(wyniki)
		log.Printf("CACHE - save JSON to cache - key: %s\n", user )
		c.Set(user,string(jsonwynik), cache.DefaultExpiration)
		return string(jsonwynik), nil
	}
}
//########################### GLOBAL ###########################################

var conf Configuration
var c *cache.Cache

//##############################################################################
func main() {
	//read config file---------------------------------------------------------------------
    config_file, _ := os.Open("config.json")
	defer config_file.Close()
	decoder := json.NewDecoder(config_file)
	err := decoder.Decode(&conf)
	if err != nil {
  	fmt.Println("error:", err)
	}
	//variables:
	//	conf.GoogleScope
	//  conf.ServiceAccountFile
	//	conf.SubjectAccount
	//	conf.LogFile
	//	conf.isBase64Encrypt

	//initialize cache ---------------------------------------------------------------------
	c = cache.New(5*time.Minute, 10*time.Minute)


	// open log file ---------------------------------------------------------------------
	f, err := os.OpenFile(conf.LogFile, os.O_RDWR | os.O_CREATE | os.O_APPEND, 0666)
	if err != nil {
    	log.Fatalf("error opening file: %v", err)
	}
	log.SetOutput(f)

	// request count---------------------------------------------------------------------
	counter := 0

	reader := bufio.NewReader(os.Stdin)

	for {
		counter += 1
		input, _ := reader.ReadString('\n')
		log.Println("############# request: ",counter, " ################")
		if strings.HasPrefix(input, "auth") {
			s := strings.Split(input, "#")
		 // s[0]="auth"; s[1]=user; s[2]=authgroups
			if len(s)==3 {
						log.Println("AUTH  - checking auth, user:(",s[1],") groups:",s[2])

	    			answer, err := check_access_by_member(s[1],s[2])
						if err != nil {
  						fmt.Println("error:", err)
						}
						log.Println("AUTH  - STDOUT: ", answer)
						fmt.Println(answer)
			}else{
				log.Println("AUTH  - Bad request, return NULL")
				fmt.Println("NULL")
			}
			continue
		}

		if strings.HasPrefix(input, "json") {
			s := strings.Split(input, "#")
			//s[0]="json"; s[1]=user;
			if len(s)==2 {
						log.Println("JSON  - fetch json, user:", s[1])

	    				answer, err := get_json(s[1])
						if err != nil {
  							log.Println("JSON  - error:", err)
  							fmt.Println("NULL")
  							continue
						}
						if conf.Base64Encrypt   {
							log.Println("JSON  - STDOUT: ", base64.StdEncoding.EncodeToString([]byte(answer)))
							fmt.Println(base64.StdEncoding.EncodeToString([]byte(answer)))
						}else {
							log.Println("JSON  - STDOUT: ", answer)
							fmt.Println(answer)
						}
			}else{
				log.Println("AUTH  - Bad request, return NULL")
				fmt.Println("NULL")
			}
		}else {
			log.Println("ERROR - bad request, return NULL")
			fmt.Println("NULL")
		}

	}
}
